chain/chain/src/flat_storage_resharder.rs

```diff
@@ -62,71 +52,81 @@ use std::iter;
-        sender: ReshardingSender,
-        controller: FlatStorageResharderController,
+        controller: ReshardingHandle,
         resharding_config: MutableConfigValue<ReshardingConfig>,
     ) -> Self {
-        let resharding_event = Arc::new(Mutex::new(None));
-        Self {
-            runtime,
-            resharding_event,
-            sender,
-            controller,
-            resharding_config,
-            #[cfg(feature = "test_features")]
-            adv_task_delay_by_blocks: 0,
-        }
+        Self { epoch_manager, runtime, handle: controller, resharding_config }
     }
 
-    /// Starts a resharding event.
-    ///
-    /// For now, only splitting a shard is supported.
-    ///
-    /// # Args:
-    /// * `event_type`: the type of resharding event
-    /// * `shard_layout`: the new shard layout
-    pub fn start_resharding(
-        &self,
-        event_type: ReshardingEventType,
-        shard_layout: &ShardLayout,
-    ) -> Result<(), Error> {
-        match event_type {
-            ReshardingEventType::SplitShard(params) => self.split_shard(params, shard_layout),
+    /// Main function to start resharding. This function is a long running cancellable task.
+    /// This is called from the resharding actor and is blocking till the resharding of flat storage
+    /// is completed.
+    pub fn start_resharding(&self, event: &ReshardingSplitShardParams) -> Result<(), Error> {
+        let status = self.runtime.store().flat_store().get_flat_storage_status(event.parent_shard);
+        let Ok(FlatStorageStatus::Ready(FlatStorageReadyStatus { flat_head })) = status else {
+            error!(target: "resharding", ?status, ?event, "flat storage shard split task: parent shard is not ready");
+            panic!("impossible to recover from a flat storage split shard failure!");
+        };
+
+        let next_epoch_id =
+            self.epoch_manager.get_next_epoch_id_from_prev_block(&event.resharding_block.hash)?;
+        let shard_layout = self.epoch_manager.get_shard_layout(&next_epoch_id)?;
+
+        let split_params = ParentSplitParameters {
+            left_child_shard: event.left_child_shard,
+            right_child_shard: event.right_child_shard,
+            shard_layout,
+            resharding_blocks: vec![event.resharding_block],
+            flat_head,
+        };
+        self.start_resharding_impl(event.parent_shard, split_params);
+
+        Ok(())
+    }
+
+    fn start_resharding_impl(&self, parent_shard: ShardUId, split_params: ParentSplitParameters) {
+        let left_child_shard = split_params.left_child_shard;
+        let right_child_shard = split_params.right_child_shard;
+        match self.split_shard_task(parent_shard, split_params) {
+            // All good.
+            FlatStorageReshardingTaskResult::Successful { .. } => {}
+            FlatStorageReshardingTaskResult::Failed => {
+                panic!("impossible to recover from a flat storage split shard failure!")
+            }
```

Comment:
not your PR but would be nice to provide an error message

Reply:
mini nit: Would be nice to order the variants by how common they are - Successful is what we expect most of the times, Cancelled rarely and Failed almost never.